Enable execution on Mac OS X

jmstover opened the pull request:
This enables us to execute Singularity containers on Mac OS X.

This provides the needed Singularity code to run a container on Mac OS X. We are still limited in functionality, but with this we can accomplish:

```
singularity run lolcow.sif
singularity exec lolcow.sif pwd
singularity shell lolcow.sif
```

The users $HOME is _automatically_ made available to the container.

Finally...
Fixes: #109 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)


Attn: @singularity-maintainers

ctmadison committed:
initial merge w/2575

jmstover committed:
startvm_linux.go

Cleanup Linux VM code.

ctmadison committed:
cleaned up serial console code

jmstover committed:
Additional CLI options

Add --vm-ram, --vm-cpu, and --syos CLI options

ctmadison committed:
made more consistant with linux version

jmstover committed:
Bind mount processing

Start processing of User Bind Mounts
Hard code a $HOME bind mount

jmstover committed:
Move from internal/app/singularity -> cmd/internal/cli

Add constant $HOME mount from Host

jmstover committed:
Change isSyOS -> IsSyOS

jmstover committed:
Pass Binds as a Kernel argument

Change location of xhyve binary to:
/usr/local/libexec/singularity/vm/xhyve

jmstover committed:
Update error message when kernel or initramfs do not exist.

jmstover committed:
Add --vm-err option to attach os.Stderr to command

- Only attach command stderr to shell stderr when this flag is
given or we are running with debug

- Fix Darwin stdin/stdout/stderr connections

jmstover committed:
Fix Copyright date range

jmstover committed:
It really does help when you use the correct format ...

  Bus:Slot:Function

Yeah....

jmstover committed:
Get working gofmt version

jmstover committed:
Make --syos hidden

jmstover committed:
go lint can leave

jmstover committed:
Modify argv[0] of xhyve execution

jmstover committed:
Pass image name to VM

jmstover committed:
Block off unusable flags

jmstover committed:
Add --no-cleanup flag to pull command. And go fmt ... gotta do go fmt ...

jmstover committed:
Pass our current user information along

Commit:
Add support for platform-specific default values

action_flags.go contains a bunch of flags shared among several commands.
With the new VM support, using a VM should be enabled by default on
darwin and disabled by default on Linux. Move that difference to
build-time by introducing a new function initPlatformDefaults which
should be used to set any platform-specific action flags.

The list of action flags is also platform-specific, so move the list of
supported flags (not the flag definitions) to a target-specific file,
too.

Use cobra's built-in method to detect whether the user passed a flag or
not in order to implement dependent flags (--vm-cpu implicitly enables
--vm). This avoids hard-coding the default values in multiple locations.

TODO: further clean-up of the actionFlags definition?

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Fix panic in execStarter

Not all commands using execStarter accept the same set of flags. In
particular, instance start takes a reduced set of flags.

Fetch the flag definition and verify that we actually got something
before trying to check if the default value was changed by the user.

Signed-off-by: Marcelo E. Magallon <mem@river>

Commit:
Hide --syos at the place of definition

--syos is being hidden (for now?). Mark it as such where it's defined,
not where it's being added to the shell command. In this way anything
else that might use that flag will get the same behavior.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Move handling of VM options to a common location

Add a vmStarter function to be called from the platform-dependent
execStarter function, which handles all the VM options. On MacOS the
--vm option is enabled by default, so this will have the correct
behavior.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Log command used to start VM

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

jmstover commented on cmd/internal/cli/actions.go:
```diff
@@ -185,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
 	return
 }
 
+// TODO: Let's stick this in another file so that that CLI is just CLI
```
You wouldn't have access to the `IsSyOS` or `VM` variables if you moved it outside the CLI package would you?

bauerm97 replied:
Well two things here:

1. Technically you're correct here, you will lose access to package-wide variables when moving a function outside of that package. However, in its current state, `vmStarter` takes `cmd *cobra.Command` as an argument which *does* give access to that flag's value.
2. We probably want to move this function to `internal/app/singularity`, and remove `cmd *cobra.Command` as an argument. Ideally, we want to parse the CLI options which have been set in `cmd/internal/cli` into some well-defined set of arguments (see the capability code for an example of this design pattern).

mem replied:
What I was thinking was creating a struct with the configuration from the command line, and passing that to a command that needs access to to that information. So, instead of having a bunch of global variables, there would be a single one that could be passed around as needed.

bauerm97 replied:
@mem I think that's a great goal to address in another PR at some point down the line

Commit:
Tidy up darwin's startVM

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

mem commented on cmd/internal/cli/startvm_darwin.go:
```diff
@@ -0,0 +1,129 @@
+	// Bind mounts
+	singBinds := []string{""}
+
+	slot := 5
```
I can't find documentation for this. Is there something special about PCI slot 5 on xhyve?

jmstover replied:
No. I just chose `5` as the starting slot for the bind points

mem commented on cmd/internal/cli/startvm_darwin.go:
```diff
@@ -0,0 +1,129 @@
+	// Force $HOME to be mounted
+	// TODO: engineConfig.GetHomeSource() / GetHomeDest() -- should probably be used
+	homeSrc := usr.HomeDir
+	pciArgs := fmt.Sprintf("4:0,virtio-9p,home=%s", homeSrc)
```
same thing here, is there something special about PCI slot 4?

jmstover replied:
Nope. It was me choosing `4` for the `$HOME` bind coming in.

mem commented on cmd/internal/cli/startvm_darwin.go:
```diff
@@ -0,0 +1,129 @@
+	userInfo := fmt.Sprintf("%s:%s:%s", usr.Username, usr.Uid, usr.Gid)
+
+	if IsSyOS {
+		cliExtra = "syos"
```
I want to make sure this is intentional: this is discarding cliExtra passed as an argument to startVM and replacing it with syos. If that's the case, a comment would help.

jmstover replied:
Yup ... You can replicate the `--syos` flag with:  `singularity exec [image] syos` ...  We're ignoring anything passed since we want a SyOS shell ... We aren't going into the image automatically here.

Commit:
Tidy up startVM for all platforms

Factor out the common code to startVM in startvm.go. Keep the
platform-specific details (hypervisors, argument construction) in the
platform-specific getHypervisorArgs function.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -217,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
 }
 
+// TODO: Let's stick this in another file so that that CLI is just CLI
+func vmStarter(cmd *cobra.Command, image string, args []string) bool {
```
I'm not a huge fan of this design pattern of having a bool function which will sometimes return, but will other times never return.

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -217,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
+	}
+
+	if VM {
+		prepareVM(cmd, image, args)
```
Let's remove the call to `prepareVM` from this function

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -217,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
 
+// TODO: Let's stick this in another file so that that CLI is just CLI
+func vmStarter(cmd *cobra.Command, image string, args []string) bool {
+	// check if --vm-ram or --vm-cpu changed from default value
```
I think all of this code which is simply responsible for setting the value of `VM` should just live in it's own function. We should just be able to set the value of the flag. Then, back outside the function we'll be able to just use `is VM {...}`

mem replied:
Got it, done

bauerm97 commented on cmd/internal/cli/startvm.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+func prepareVM(cmd *cobra.Command, image string, args []string) {
```
I think this function can be renamed `execVM()` to better align with `execStarter()`. Also, let's move this into `actions.go`, since it's common across architectures.

bauerm97 commented on cmd/internal/cli/actions_linux.go:
```diff
@@ -34,6 +34,10 @@ import (
 
 // TODO: Let's stick this in another file so that that CLI is just CLI
 func execStarter(cobraCmd *cobra.Command, image string, args []string, name string) {
+	if vmStarter(cobraCmd, image, args) {
```
Let's remove this code. Leave `execStarter()` function as a function which is solely responsible for running the `starter` binary on linux platform. Likewise, `execVM()` can be responsible for starting the VM.

mem replied:
Let me give this a try

bauerm97 commented on cmd/internal/cli/actions_darwin.go:
```diff
@@ -11,4 +11,5 @@ import (
 
 // TODO: Let's stick this in another file so that that CLI is just CLI
 func execStarter(cobraCmd *cobra.Command, image string, args []string, name string) {
```
I would say, move the function `execStarter` here into `actions_unsupported.go` and have it be a `panic`.

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -217,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
+}
+
 // ExecCmd represents the exec command
 var ExecCmd = &cobra.Command{
```
Following from the other comments, I think that the run/exec/shell Command variables should all have their `Run` fields be a function like this:

```
func ActionRun(cmd *cobra.Command, args []string) {
	a := append([]string{"/.singularity.d/actions/exec"}, args[1:]...)
	for _, flagName := range []string{"vm-ram", "vm-cpu"} {
		if flag := cmd.Flag(flagName); flag != nil && flag.Changed {
			// the user passed an option related to the VM...
			if !VM {
				// ... but didn't enable the VM itself
				cmd.Flag("vm").Value.Set("true")
			}
			break
		}
	}

	// since --syos is a boolean, it cannot be added to the above list
	if IsSyOS && !VM {
		sylog.Warningf("The --syos option requires a virtual machine, automatically enabling --vm option.")
		cmd.Flag("vm").Value.Set("true")
	}

	if VM {
		execVM()
	}

	execStarter()
}
```

On both `darwin` && `linux`, `execVM` will be defined (in fact they will both be defined here in `actions.go`). However, on all other platforms `execStarter` will be defined as `panic()`.

bauerm97 commented on cmd/internal/cli/startvm_darwin.go:
```diff
@@ -0,0 +1,129 @@
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+func startVM(sifImage, singAction, cliExtra string, isInternal bool) error {
```
Eventually `startVM` [for all archs] should be moved to `internal/app/singularity`, but for now this is fine.

Commit:
Reorganize execStarter / execVM

Move execVM out of execStarter. Actions wishing to support VM operation
must explicitly check whether the user is requesting that and execute
the VM themselves.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -225,6 +219,10 @@ var ExecCmd = &cobra.Command{
 	PreRun:                replaceURIWithImage,
 	Run: func(cmd *cobra.Command, args []string) {
 		a := append([]string{"/.singularity.d/actions/exec"}, args[1:]...)
+		if wantsVM(cmd) {
```
Let's move this into its own function. As noted previously, just create `func ActionRun` and then here do: 

```
Run: ActionRun
```

Same comment applies for all of these related commands

bauerm97 replied:
Actually, this is kinda dumb, I don't think that makes sense here.

bauerm97 commented on cmd/internal/cli/actions_darwin.go:
```diff
@@ -11,4 +11,5 @@ import (
 
 // TODO: Let's stick this in another file so that that CLI is just CLI
 func execStarter(cobraCmd *cobra.Command, image string, args []string, name string) {
+	panic("starter is unsupported on this platform")
```
Change this file name to `actions_unsupported`, to handle windows/etc.. as well

WestleyK commented on cmd/internal/cli/actions.go:
```diff
@@ -35,48 +35,15 @@ func init() {
 	// TODO : the next n lines of code are repeating too much but I don't
 	// know how to shorten them tonight
```
I think this `TODO :` is solved, it can probably be removed.

Commit:
Enable --vm option if required by other options

Passing some options (--vm-ram, --vm-cpu, --syos) will automatically
enable the --vm option, as it does not make sense to use those flags
without a VM.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Remove resolved TODO

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Provide a common execStarter implementation for unsupported platforms

Right now only Linux supports starting containers, other platforms use a
VM where Linux is run in order to start the container. Instead of having
one file per unsupported platform, provide a common implementation for
all of them. If another platform becomes supported, it should be added
to the build tags at the top of actions_unsupported.go.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

mem committed:
Merge branch 'master' into testpr

bauerm97 committed:
Merge branch 'master' into testpr
